uniffi_bindgen/src/bindings/python/templates/RecordTemplate.py

```diff
@@ -6,7 +6,7 @@ class {{ type_name }}:
 
     {%- if rec.has_fields() %}
     @typing.no_type_check
-    def __init__(self, {% for field in rec.fields() %}
+    def __init__(self, *, {% for field in rec.fields() %}
```

Comment:
I thought about that as well. And sure, it would be possible to build that.
I'm a bit concerned about the implications thought. Suddenly adding a different argument can change how a function needs to be called. Is this too hidden of a feature? (tbf, I haven't thought enough about where I stand on that question)

Reply:
Good point: As a meta comment though, adding a different argument *without a default* necessarily means all call-sites need to change anyway, in every language. Adding a different argument with a default meaning they *might* also change in some cases for Python seems reasonable.

But we could take this even further: if there's a non-default arg after a default-arg, then we could *reorder* the args - `i32 default_first(optional i32 left = 1, i32 right);` becomes `def foo(right, *, left = 1):`. This still allows you to choose to specify named args in any order but has the following additional benefits:
* zero change to existing compatible records and functions.
* adding a new arg then has the exact same implications as other languages - non-default value means you *must* change all call-sites, but new default-arg means you need not change any call sites, even if you add that arg *before* other args with defaults.

Quite possibly there are bad implications I haven't thought about though :) Note I've no horse in this race, so these are just ideas. I'm more than happy for you and @jeddai to agree on semantics.

FWIW though, the semantics seem OK to me:
```python
>>> def default_first(right, *, left = 1):
...     print(left, right)
... 
>>> default_first(left=1, right=2)
1 2
>>> # All positional args are args without defaults in the same order they are declared.
>>> default_first(2)
1 2
>>> # Full keyword support exists even for args without a default
>>> default_first(right=2)
1 2
>>> 
```

Reply:
So I did implement this now and I have some thoughts:

* For records this works. There you pass multiple things to the initializer and they have named fields and thus order is less important. However I then prefer enforcing naming, simply to avoid ordering issues.
* For functions it feels weird. If your udl defines `(optional i32 left = 1, i32 right)` this suddenly becomes `(right, left=1)` in the calling code. IMO this is confusing, suddenly looking at the definition and at the call site doesn't match up anymore. I personally would say we do not touch that case and do not support defaulted values before not-defaulted ones in functions.

All in all I prefer using only the first 2 commits here, the struct named parameter handling.
And drop the third one, the one for functions.

Reply:
> I personally would say we do not touch that case and do not support defaulted values before not-defaulted ones in functions.

Is there a reason not to support the simple `fn foo(self, *, ...)` syntax for such functions? IOW, don't do the fancy arg repositioning but instead force such functions to always used named args for all args? IIUC that will be a breaking change, but that seems like a reasonable tradeoff if the alternative is such functions simply don't work at all.

Reply:
Yeah I'm on board with this syntax, the more explicit we are when using constructors the better imo.